pkg/networkpolicy/packet.go

```diff
@@ -29,59 +33,110 @@ func (p packet) String() string {
 func parsePacket(b []byte) (packet, error) {
 	t := packet{}
-	if b == nil {
-		return t, fmt.Errorf("empty payload")
+	if len(b) < 20 {
+		// 20 is the minimum length of an IPv4 header (IPv6 is 40)
+		return t, ErrorTooShort
 	}
 	version := int(b[0] >> 4)
 	// initialize variables
-	var hdrlen, protocol int
+	var protocol, l4offset, nxtHeader, payloadOffset int
 	switch version {
 	case 4:
 		t.family = v1.IPv4Protocol
-		hdrlen = int(b[0]&0x0f) << 2
-		if len(b) < hdrlen+4 {
-			return t, fmt.Errorf("payload to short, received %d expected at least %d", len(b), hdrlen+4)
+		hdrlen := int(b[0]&0x0f) * 4 // (header length in 32-bit words)
+		if hdrlen < 20 {
+			return t, ErrorCorrupted
+		}
+		l4offset = hdrlen
+		if l4offset >= len(b) {
+			return t, ErrorTooShort
 		}
 		t.srcIP = net.IPv4(b[12], b[13], b[14], b[15])
 		t.dstIP = net.IPv4(b[16], b[17], b[18], b[19])
 		protocol = int(b[9])
+		// IPv4 fragments:
+		// Since the conntracker is always used in K8s, IPv4 fragments
+		// will never be passed via the nfqueue. Packets are
+		// re-assembled by the kernel. Please see:
+		// https://unix.stackexchange.com/questions/650790/unwanted-defragmentation-of-forwarded-ipv4-packets
 	case 6:
 		t.family = v1.IPv6Protocol
-		hdrlen = 40
-		if len(b) < hdrlen+4 {
-			return t, fmt.Errorf("payload to short, received %d expected at least %d", len(b), hdrlen+4)
+		if len(b) < 48 {
+			// 40 is the minimum length of an IPv6 header, and 8 is
+			// the minimum lenght of an extension or L4 header
+			return t, ErrorTooShort
 		}
 		t.srcIP = make(net.IP, net.IPv6len)
 		copy(t.srcIP, b[8:24])
 		t.dstIP = make(net.IP, net.IPv6len)
 		copy(t.dstIP, b[24:40])
-		// NextHeader (not extension headers supported)
-		protocol = int(b[6])
+		// Handle extension headers.
+		nxtHeader = int(b[6])
+		l4offset = 40
+		for nxtHeader == syscall.IPPROTO_DSTOPTS || nxtHeader == syscall.IPPROTO_HOPOPTS || nxtHeader == syscall.IPPROTO_ROUTING {
+			// These headers have a lenght in 8-octet units, not
+			// including the first 8 octets
+			nxtHeader = int(b[l4offset])
+			l4offset += (8 + int(b[l4offset+1])*8)
+			// Now l4offset points to either another extension header,
+			// or an L4 header. So we must have at least 8 byte data
+			// after this (minimum extension header size)
+			if (l4offset + 8) >= len(b) {
+				return t, ErrorTooShort
+			}
+		}
+		if nxtHeader == syscall.IPPROTO_FRAGMENT {
+			// Only the first fragment has the L4 header
+			fragOffset := int(binary.BigEndian.Uint16(b[l4offset+2 : l4offset+4]))
+			if fragOffset&0xfff8 == 0 {
+				nxtHeader = int(b[l4offset])
+				l4offset += 8
+				// Here it's assumed that the fragment is the last
+				// extension header before the L4 header. But this is
+				// only a *recommendation*.
+				// TODO: handle extension headers in non-recommended order
+			} else {
+				// Not first fragment, we have no L4 header and the
+				// payload begins after this header
+				l4offset = -1
```

Comment:
is `protocol = nxtHeader` correct in this case? it will be `syscall.IPPROTO_FRAGMENT ` 

`l4offset = -1`. does not seem to be handled later, should this return here?

Reply:
> is protocol = nxtHeader correct in this case? it will be syscall.IPPROTO_FRAGMENT

It's intentional, but strictly not correct since there is an L4 protocol in the nxtHeader field. _But there is no L4 header!_ The fragment data begins immediately after the fragment header. So to identify the packet as a fragment, syscall.IPPROTO_FRAGMENT is kept in the `protocol` variable. I'll make a comment.

Reply:
> l4offset = -1. does not seem to be handled later, should this return here?

It's probably a just-in-case left-over. I'll check it.